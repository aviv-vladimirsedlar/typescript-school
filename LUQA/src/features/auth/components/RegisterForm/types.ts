export interface FormFields {
  email: string;
  password: string;
  passwordConfirm: string;
  firstName: string;
  lastName: string;
}
