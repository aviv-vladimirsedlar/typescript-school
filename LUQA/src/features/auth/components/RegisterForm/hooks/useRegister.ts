import { useMutation } from "@tanstack/react-query";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";

import { registerRequest } from "../../../auth.api";
import { loginSuccess } from "../../../auth.slice";

export const useRegister = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const mutation = useMutation({
    mutationFn: registerRequest,
    onSuccess: (data) => {
      // Dispatch Redux action on successful registration
      dispatch(loginSuccess({ user: data.user }));
      // Navigate to the home page
      navigate("/");
    },
    onError: (error) => {
      // Handle registration error
      console.error("Registration failed:", error);
    },
  });

  return mutation;
};
