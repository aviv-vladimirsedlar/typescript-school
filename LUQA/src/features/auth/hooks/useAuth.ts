import { useSelector } from "react-redux";

import { RootState } from "../../../config/store";

export const useAuth = () => {
  const isAuthenticated = useSelector(
    (state: RootState) => state.auth.isAuthenticated
  );

  const user = useSelector((state: RootState) => {
    return state.auth.user;
  });

  return { isAuthenticated, user };
};
