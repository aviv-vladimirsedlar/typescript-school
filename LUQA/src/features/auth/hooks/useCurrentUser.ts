import { useQuery } from "@tanstack/react-query";
import { useMemo } from "react";
import { useDispatch, useSelector } from "react-redux";

import axiosInstance from "../../../config/api";
import { RootState } from "../../../config/store";
import { loginSuccess } from "../auth.slice";

const getCurrentUser = async () => {
  const response = await axiosInstance.get("/users/me");
  return response.data;
};

export const useCurrentUser = () => {
  const dispatch = useDispatch();

  const isAuthenticated = useSelector(
    (state: RootState) => state.auth.isAuthenticated
  );
  const currentUser = useSelector((state: RootState) => state.auth.user);

  const isAdmin = useMemo(
    () =>
      currentUser?.roles?.some((userRole) => userRole.role.name === "admin"),
    [currentUser]
  );
  const isAuthor = useMemo(
    () =>
      currentUser?.roles?.some((userRole) => userRole.role.name === "author"),
    [currentUser]
  );

  const { data, isLoading, error } = useQuery({
    queryKey: ["currentUser"], // Use array syntax for query keys
    queryFn: async () => {
      const user = await getCurrentUser();
      // Handle onSuccess logic here
      dispatch(loginSuccess({ user }));
      return user;
    },
    retry: false,
    enabled: isAuthenticated,
  });

  // Handle errors directly
  if (error) {
    console.error("Login error:", error);
  }

  return { currentUser: data, isAdmin, isAuthor, isLoading };
};
