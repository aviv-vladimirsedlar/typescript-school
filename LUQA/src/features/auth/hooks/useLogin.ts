import { useMutation } from "@tanstack/react-query";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";

import { loginRequest } from "../auth.api";
import { loginSuccess } from "../auth.slice";

export const useLogin = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const mutation = useMutation({
    mutationFn: loginRequest,
    onSuccess: (data) => {
      dispatch(loginSuccess({ user: data.user })); // Dispatch Redux action
      navigate("/"); // Navigate to the home page
    },
    onError: (error) => {
      console.error("Login failed:", error); // Handle error
    },
  });
  console.log(mutation);

  return mutation;
};
