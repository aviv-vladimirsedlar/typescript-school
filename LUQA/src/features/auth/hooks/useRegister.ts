import { useMutation } from "react-query";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";

import axiosInstance from "../../../config/api";
import { loginSuccess } from "../auth.slice";

interface RegisterData {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
}

const registerRequest = async (data: RegisterData) => {
  const response = await axiosInstance.post("/users/register", data);
  return response.data;
};

export const useRegister = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  return useMutation(registerRequest, {
    onSuccess: (data) => {
      dispatch(loginSuccess({ user: data.user }));
      navigate("/");
    },
  });
};
