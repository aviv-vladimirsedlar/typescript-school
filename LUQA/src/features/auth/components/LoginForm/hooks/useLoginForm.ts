import { zodResolver } from "@hookform/resolvers/zod";
import { useState } from "react";
import { useForm, SubmitHandler } from "react-hook-form";
import { z } from "zod";

import { useLogin } from "./useLogin"; // Assuming you have a login mutation hook

// Define the validation schema with Zod
const loginSchema = z.object({
  email: z
    .string()
    .email("Invalid email address")
    .nonempty("Email is required"),
  password: z
    .string()
    .min(6, "Password must be at least 6 characters")
    .nonempty("Password is required"),
});

// Infer the form type from the Zod schema
type LoginFormInputs = z.infer<typeof loginSchema>;

// Define the error type based on your backend's error structure
interface LoginError {
  response?: {
    data?: {
      message?: string;
    };
  };
}

export const useLoginForm = () => {
  const [errorMessage, setErrorMessage] = useState<string>("");

  // Set up react-hook-form with Zod validation
  const {
    control,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<LoginFormInputs>({
    resolver: zodResolver(loginSchema),
  });

  // Hook for login mutation
  const { mutate: login } = useLogin();

  // Form submission handler
  const onSubmit: SubmitHandler<LoginFormInputs> = (data) => {
    login(data, {
      onError: (error: unknown) => {
        // Explicitly cast error to LoginError
        const typedError = error as LoginError;
        const message =
          typedError?.response?.data?.message || "An error occurred";
        setErrorMessage(message);
      },
    });
  };

  return {
    control, // For use with Controller in the LoginForm component
    handleSubmit, // To wrap the onSubmit handler
    onSubmit, // Custom submit handler
    isLoading: isSubmitting, // To show loading states
    errors, // Validation errors
    errorMessage, // General error message for display
  };
};
