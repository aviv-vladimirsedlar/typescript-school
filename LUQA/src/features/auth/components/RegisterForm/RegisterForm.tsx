import { Box } from "@gemini/core";
import { Button, TextField } from "@gemini/ui";
import React from "react";
import { Controller } from "react-hook-form";

import { useRegisterForm } from "./hooks/useRegisterForm";

export const RegisterForm: React.FC = () => {
  const { control, handleSubmit, onSubmit, isLoading, errors, errorMessage } =
    useRegisterForm();

  return (
    <Box
      as="form"
      onSubmit={handleSubmit(onSubmit)}
      paddingVertical="spacing.20"
      paddingHorizontal="spacing.20"
    >
      <Box width="100%" marginBottom="spacing.12">
        <Box
          display="grid"
          marginVertical="spacing.24"
          gap="spacing.16"
          width="100%"
          gridTemplateColumns={{
            "breakpoint.sm": "repeat(1, 1fr)",
            "breakpoint.lg": "repeat(2, minmax(0, 1fr))",
          }}
        >
          <Controller
            name="firstName"
            control={control}
            render={({ field }) => (
              <TextField
                {...field}
                testId="first-name-input"
                label="First name"
                placeholder="Enter your first name"
                error={
                  typeof errors.firstName?.message === "string"
                    ? errors.firstName.message
                    : undefined
                }
                isOptional={false}
              />
            )}
          />
          <Controller
            name="lastName"
            control={control}
            render={({ field }) => (
              <TextField
                {...field}
                testId="last-name-input"
                label="Last name"
                placeholder="Enter your last name"
                error={
                  typeof errors.lastName?.message === "string"
                    ? errors.lastName.message
                    : undefined
                }
                isOptional={false}
              />
            )}
          />
        </Box>
      </Box>
      <Box width="100%" marginBottom="spacing.12">
        <Controller
          name="email"
          control={control}
          render={({ field }) => (
            <TextField
              {...field}
              testId="email-input"
              label="Your email"
              placeholder="Enter your email"
              error={
                typeof errors.email?.message === "string"
                  ? errors.email.message
                  : undefined
              }
              isOptional={false}
            />
          )}
        />
      </Box>
      <Box width="100%" marginBottom="spacing.12">
        <Controller
          name="password"
          control={control}
          render={({ field }) => (
            <TextField
              {...field}
              testId="password-input"
              label="Password"
              placeholder="Enter your password"
              type="password"
              error={
                typeof errors.password?.message === "string"
                  ? errors.password.message
                  : undefined
              }
              isOptional={false}
            />
          )}
        />
      </Box>
      <Box width="100%" marginBottom="spacing.12">
        <Controller
          name="passwordConfirm"
          control={control}
          render={({ field }) => (
            <TextField
              {...field}
              testId="confirm-password-input"
              label="Confirm password"
              placeholder="Re-enter your password"
              type="password"
              error={
                typeof errors.passwordConfirm?.message === "string"
                  ? errors.passwordConfirm.message
                  : undefined
              }
              isOptional={false}
            />
          )}
        />
      </Box>

      <Button
        testId="btn-register"
        isLoading={isLoading}
        type="submit"
        size="48"
      >
        Submit
      </Button>

      {!!errorMessage && (
        <Box
          testId="register-error-message"
          display="flex"
          gap="spacing.20"
          backgroundColor="color.border.status.error.default"
          marginTop="spacing.16"
          typography="typography.body.14.bold"
          color="color.content.constant.white"
          textAlign="left"
          padding="spacing.12"
          paddingLeft="spacing.16"
          zIndex="zIndex.4"
        >
          {errorMessage}
        </Box>
      )}
    </Box>
  );
};
