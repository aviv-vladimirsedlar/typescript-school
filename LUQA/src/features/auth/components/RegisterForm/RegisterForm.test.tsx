import { GeminiProvider } from "@gemini/core";
import { AVIV_THEME } from "@gemini/tokens";
import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from "@testing-library/react";
import React from "react";

import "@testing-library/jest-dom";
import { findInputAndChangeText } from "../../../../common/utils/test.util";
import { useRegister } from "./hooks/useRegister";

import { RegisterForm } from "./RegisterForm";

jest.mock("../../../common/hooks/useRegister");

const mockMutate = jest.fn();
const mockUseRegister = useRegister as jest.MockedFunction<typeof useRegister>;

const renderComponent = () => (
  <GeminiProvider theme={AVIV_THEME}>
    <RegisterForm />
  </GeminiProvider>
);

describe("RegisterForm", () => {
  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();
    mockUseRegister.mockReturnValue({
      mutate: mockMutate,
      isLoading: false,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);
  });

  it("should match snapshot when initially rendered", () => {
    const { asFragment } = render(renderComponent());
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the form and handle successful registration", async () => {
    mockMutate.mockImplementation(
      (data, { onSuccess }) => onSuccess && onSuccess()
    );
    render(renderComponent());
    await act(async () => {
      findInputAndChangeText("first-name-input", "John");
      findInputAndChangeText("last-name-input", "Doe");
      findInputAndChangeText("email-input", "john.doe@example.com");
      findInputAndChangeText("password-input", "Password123");
      findInputAndChangeText("confirm-password-input", "Password123");
      fireEvent.click(screen.getByTestId("btn-register"));
    });
    await waitFor(() => expect(mockMutate).toHaveBeenCalled());
    expect(
      screen.queryByText(/An unknown error occurred/i)
    ).not.toBeInTheDocument();
  });

  it("should show error message on failed registration", async () => {
    mockMutate.mockImplementation(
      (data, { onError }) =>
        onError &&
        onError({ response: { data: { message: "Registration failed" } } })
    );
    render(renderComponent());
    await act(async () => {
      findInputAndChangeText("first-name-input", "John");
      findInputAndChangeText("last-name-input", "Doe");
      findInputAndChangeText("email-input", "john.doe@example.com");
      findInputAndChangeText("password-input", "Password123");
      findInputAndChangeText("confirm-password-input", "Password123");
      fireEvent.click(screen.getByTestId("btn-register"));
    });
    await waitFor(() =>
      expect(screen.getByText(/Registration failed/i)).toBeInTheDocument()
    );
  });

  it("should validate required fields and show validation errors", async () => {
    render(renderComponent());
    await act(async () => {
      fireEvent.click(screen.getByTestId("btn-register"));
    });
    await waitFor(() => {
      expect(screen.getByTestId("first-name-input")).toHaveTextContent(
        "is required"
      );
      expect(screen.getByTestId("last-name-input")).toHaveTextContent(
        "is required"
      );
      expect(screen.getByTestId("email-input")).toHaveTextContent(
        "is required"
      );
      expect(screen.getByTestId("password-input")).toHaveTextContent(
        "is required"
      );
      expect(screen.getByTestId("confirm-password-input")).toHaveTextContent(
        "is required"
      );
    });
  });
});
