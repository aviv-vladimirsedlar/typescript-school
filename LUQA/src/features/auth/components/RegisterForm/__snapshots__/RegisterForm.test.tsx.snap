// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm should match snapshot when initially rendered 1`] = `
<DocumentFragment>
  <form
    class="mx-auto w-full max-w-2xl"
  >
    <div
      class="css-167vhyo"
    >
      <div
        class="css-mbjr46"
      >
        <div
          class="css-ow2jnx"
          data-testid="first-name-input"
        >
          <div
            class="css-j7qwjs"
          >
            <div
              class="css-6xq2vb"
            >
              <label
                class="css-1azm4t"
                for="react-aria-:r0:"
                id="react-aria-:r1:"
              >
                First name
                <span
                  class="css-uv7gtw"
                >
                  *
                </span>
              </label>
            </div>
          </div>
          <div
            class="css-1c1vv48"
          >
            <input
              aria-labelledby="react-aria-:r1:"
              aria-required="true"
              class="css-pcexyr"
              id="react-aria-:r0:"
              inputmode="text"
              size="1"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="css-ow2jnx"
          data-testid="last-name-input"
        >
          <div
            class="css-j7qwjs"
          >
            <div
              class="css-6xq2vb"
            >
              <label
                class="css-1azm4t"
                for="react-aria-:r5:"
                id="react-aria-:r6:"
              >
                Last name
                <span
                  class="css-uv7gtw"
                >
                  *
                </span>
              </label>
            </div>
          </div>
          <div
            class="css-1c1vv48"
          >
            <input
              aria-labelledby="react-aria-:r6:"
              aria-required="true"
              class="css-pcexyr"
              id="react-aria-:r5:"
              inputmode="text"
              size="1"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-167vhyo"
    >
      <div
        class="css-ow2jnx"
        data-testid="email-input"
      >
        <div
          class="css-j7qwjs"
        >
          <div
            class="css-6xq2vb"
          >
            <label
              class="css-1azm4t"
              for="react-aria-:ra:"
              id="react-aria-:rb:"
            >
              Your email
              <span
                class="css-uv7gtw"
              >
                *
              </span>
            </label>
          </div>
        </div>
        <div
          class="css-1c1vv48"
        >
          <input
            aria-labelledby="react-aria-:rb:"
            aria-required="true"
            class="css-pcexyr"
            id="react-aria-:ra:"
            inputmode="text"
            size="1"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-167vhyo"
    >
      <div
        class="css-ow2jnx"
        data-testid="password-input"
      >
        <div
          class="css-j7qwjs"
        >
          <div
            class="css-6xq2vb"
          >
            <label
              class="css-1azm4t"
              for="react-aria-:rf:"
              id="react-aria-:rg:"
            >
              Password
              <span
                class="css-uv7gtw"
              >
                *
              </span>
            </label>
          </div>
        </div>
        <div
          class="css-1c1vv48"
        >
          <input
            aria-labelledby="react-aria-:rg:"
            aria-required="true"
            class="css-pcexyr"
            id="react-aria-:rf:"
            inputmode="text"
            size="1"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-167vhyo"
    >
      <div
        class="css-ow2jnx"
        data-testid="confirm-password-input"
      >
        <div
          class="css-j7qwjs"
        >
          <div
            class="css-6xq2vb"
          >
            <label
              class="css-1azm4t"
              for="react-aria-:rk:"
              id="react-aria-:rl:"
            >
              Confirm password
              <span
                class="css-uv7gtw"
              >
                *
              </span>
            </label>
          </div>
        </div>
        <div
          class="css-1c1vv48"
        >
          <input
            aria-labelledby="react-aria-:rl:"
            aria-required="true"
            class="css-pcexyr"
            id="react-aria-:rk:"
            inputmode="text"
            size="1"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <button
      class="css-1bzl4wl"
      data-testid="btn-register"
      type="button"
    >
      <div
        class="css-o395q5"
      >
        <span
          class="css-1gur7lg"
        >
          Submit
        </span>
      </div>
    </button>
  </form>
</DocumentFragment>
`;
